addon.py

```diff
@@ -1,78 +1,130 @@
-    elif path is not None:
-        files = requests.post(this_addon.getSetting('server') + '/dirparser', data={'dir': path[0], 'token': token}).json()
-        for file in files.get('contents'):
-            if(file.get('type') == 'directory'):
-                url = build_url({'mode': 'files', 'path': files.get('path') + file.get('name')})
-                item = xbmcgui.ListItem(file.get('name'), iconImage='DefaultFolder.png')
-                xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=item, isFolder=True)
+            self.add_item(name=folder['name'],
+                          url=self.build_url(
+                              mode='files',
+                              path=os.path.join(folder['path'], folder['name'])
+                          ),
+                          is_folder=True)
+        xbmcplugin.endOfDirectory(self.handle)
+
+    def play_song(self, path):
+        song_meta_url = urlparse.urljoin(self.server, 'db/metadata')
+        response = requests.post(song_meta_url, data={'filepath': path,
+                                                      'token': self.token}).json()
+        song_meta = response['metadata']
+
+        song_url = urlparse.urljoin(self.server,
+                                    'media/' + urllib.quote_plus(path))
+        item = xbmcgui.ListItem(
+            path='{}?{}'.format(song_url, urllib.urlencode({'token': self.token}))
+        )
+        item.setInfo('music', {
+            'artist': song_meta['artist'],
+            'album': song_meta['album'],
+            'tracknumber': song_meta['track'],
+            'title': song_meta['title'],
+            'year': song_meta['year']
+        })
+        xbmcplugin.setResolvedUrl(self.handle, succeeded=True, listitem=item)
+
+    def act(self, mode, path):
+        if mode is None:
+            self.build_filesystem()
+        elif mode == 'files':
+            if path is None and len(self.folders) == 1:
+                self.list_folder_contents(self.folders[0])
+            elif path is None and len(self.folders) > 1:
+                self.list_folders(self.folders)
             else:
-                url = build_url({'mode': 'play', 'path': files.get('path') + file.get('name')})
-                item = xbmcgui.ListItem(file.get('name'))
-                item.setProperty('IsPlayable', 'true')
-                xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=item, isFolder=False)
-        xbmcplugin.endOfDirectory(addon_handle)
-elif mode[0] == 'play':
-    play_song(args.get('path')[0], token=token)
\ No newline at end of file
+                self.list_folder_contents(path)
+        else:
+            self.play_song(path[0])
+
+
+if __name__ == '__main__':
+    print(sys.argv)
+
+    base_url = sys.argv[0]
+    addon_handle = int(sys.argv[1])
+    args = urlparse.parse_qs(sys.argv[2][1:])
+
+    operator = AddonOperator(addon_handle, base_url)
+    operator.act(args.get('mode'), args.get('path'))
```

Comment:
There is a problem here when I call it the first time with mode=files because there is not path=...